src/seeds.accounts.cpp

```diff
@@ -976,3 +981,151 @@ uint64_t accounts::rep_score(name user)
 
     return ritr->rank;
 }
+
+void accounts::send_punish (name account, uint64_t points) {
+  action(
+    permission_level(get_self(), "active"_n),
+    get_self(),
+    "punish"_n,
+    std::make_tuple(account, points)
+  ).send();
+}
+
+void accounts::punish (name account, uint64_t points) {
+  require_auth(get_self());
+  send_subrep(account, points);
+  pnshvouchers(account, points, uint64_t(0));
+}
+
+void accounts::pnshvouchers (name account, uint64_t points, uint64_t start) {
+  require_auth(get_self());
+
+  vouch_tables vouch(get_self(), account.value);
+
+  auto vitr = start == 0 ? vouch.begin() : vouch.find(start);
+  uint64_t count = 0;
+  uint64_t batch_size = config_get("batchsize"_n);
+  uint64_t lost_points = points * config_float_get("flag.vouch.p"_n);
+
+  while (vitr != vouch.end() && count < batch_size) {
+    send_subrep(vitr -> sponsor, lost_points);
+    vitr++;
+    count++;
+  }
+
+  if (vitr != vouch.end()) {
+    uint64_t next_value = (vitr -> sponsor).value;
+    
+    action next_execution(
+        permission_level{get_self(), "active"_n},
+        get_self(),
+        "pnshvouchers"_n,
+        std::make_tuple(account, points, next_value)
+    );
+
+    transaction tx;
+    tx.actions.emplace_back(next_execution);
+    // tx.delay_sec = 1;
+    tx.send(next_value, _self);
+  }
+}
+
+void accounts::flag (name from, name to) {
+  require_auth(from);
+
+  if (from == to) { return; }
+
+  flag_points_tables flags(get_self(), to.value);
+  flag_points_tables total_flags(get_self(), flag_total_scope.value);
+  flag_points_tables removed_flags(get_self(), flag_remove_scope.value);
+
+  auto fitr = flags.find(from.value);
+  check(fitr == flags.end(), "a user can only flag another user once");
+
+  auto to_rep = rep.find(to.value);
+  check(to_rep != rep.end(), to.to_string() + " user has no reputation");
```

Comment:
Should fail silently, not fail the transaction? If I flag someone but they don't have a reputation, the fact I flagged them is still OK.

Reply:
Ok, I added a validation. In case the flagged user does not have rep, the total_flag points are still increased, but the total_removed not, so when the user has rep and if the user is flagged again, the user will lose the pending points as well